pkg/reconciler/pipelinerun/resources/validate_params.go

```diff
@@ -90,108 +88,56 @@ func ValidateObjectParamRequiredKeys(pipelineParameters []v1beta1.ParamSpec, pip
 }
 
 // ValidateParamArrayIndex validate if the array indexing param reference target is existent
-func ValidateParamArrayIndex(ctx context.Context, p *v1beta1.PipelineSpec, pr *v1beta1.PipelineRun) error {
+func ValidateParamArrayIndex(ctx context.Context, pipelineSpec *v1beta1.PipelineSpec, prParams []v1beta1.Param) error {
 	if !config.CheckAlphaOrBetaAPIFields(ctx) {
 		return nil
 	}
 
-	arrayParams := extractParamIndexes(p.Params, pr.Spec.Params)
+	// Collect all array params lengths
+	// Example: {"$(params.array-params)": 2, "$(params[\"array-params\"])": 2, "$(params['array-params'])": 2}
+	arrayParamsLengths := taskrun.ExtractParamArrayLengths(pipelineSpec.Params, prParams)
 
-	outofBoundParams := sets.String{}
-
-	// collect all the references
-	for i := range p.Tasks {
-		findInvalidParamArrayReferences(p.Tasks[i].Params, arrayParams, &outofBoundParams)
-		if p.Tasks[i].IsMatrixed() {
-			findInvalidParamArrayReferences(p.Tasks[i].Matrix.Params, arrayParams, &outofBoundParams)
+	// extract all array indexing references, for example []{"$(params.array-params[1])"}
+	arrayIndexParamRefs := []string{}
+	for i := range pipelineSpec.Tasks {
+		extractArrayIndexingRefsFromParams(pipelineSpec.Tasks[i].Params, &arrayIndexParamRefs)
+		if pipelineSpec.Tasks[i].IsMatrixed() {
+			extractArrayIndexingRefsFromParams(pipelineSpec.Tasks[i].Matrix.Params, &arrayIndexParamRefs)
```

Comment:
thank you @Yongxuanzhang for this refactoring, this is a huge improvement on what we have 👍 

Can we maintain a list of params `[]param` and then call `extractArrayIndexingRefsFromParams` with that list instead of calling the same extract function multiple times. It's just easier to read, for example:

```
	var ps []v1beta1.Param
	for i := range pipelineSpec.Tasks {
		ps = pipelineSpec.Tasks[i].Params
		if pipelineSpec.Tasks[i].IsMatrixed() {
			ps = append(ps, pipelineSpec.Tasks[i].Matrix.Params...)
		}
	}
	extractArrayIndexingRefsFromParams(ps, &arrayIndexParamRefs)
```

Reply:
ah yes! That makes sense. I can do that